plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/AlarmManager.java

```diff
@@ -0,0 +1,263 @@
+    /**
+     * @param id The ID of a PDE Alarm.
+     * @return Data describing the alarm with the given ID if it exists, null
+     * otherwise.
+     */
+    private Alarm getAlarmData(final int id) {
+        for (final Alarm alarm : activeAlarms) {
+            if (alarm.getId() == id) {
+                return alarm;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * @return A list containing all PDE alarms.
+     */
+    public List<PdeAlarmDto> getAlarms() {
+        final List<PdeAlarmDto> result = new ArrayList<>();
+
+        final List<Alarm> allAlarms = new ArrayList<>();
+        allAlarms.addAll(activeAlarms);
+        allAlarms.addAll(clearedAlarms);
+
+        for (final Alarm alarm : allAlarms) {
+            result.add(alarm.toPdeAlarm());
+        }
+
+        return result;
+    }
+
+    /**
+     * @param cause An alarm cause.
+     * @return A list of raw data describing currently active alarms with the
+     * given alarm cause.
+     */
+    public List<Alarm> getActiveAlarmData(final AlarmCause cause) {
+        Objects.requireNonNull(cause, "Expected alarm cause.");
+        return activeAlarms.stream()
+            .filter(alarm -> cause == alarm.getCause())
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * @param causes A list of alarm causes.
+     * @return A list of raw data describing currently active alarms whose alarm
+     * cause is any of the causes provided.
+     */
+    public List<Alarm> getActiveAlarmData(final List<AlarmCause> causes) {
+        Objects.requireNonNull(causes, "Expected alarm causes.");
+        return activeAlarms.stream()
+            .filter(alarm -> causes.contains(alarm.getCause()))
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * @param id The ID of a PDE Alarm.
+     * @return The PDE Alarm with the given ID if it exists, null otherwise.
+     */
+    public PdeAlarmDto getAlarmDto(final int id) {
+        final Alarm alarmData = getAlarmData(id);
+        if (alarmData != null) {
+            return alarmData.toPdeAlarm();
+        }
+        return null;
+    }
+
+    /**
+     * @param id           ID of an alarm.
+     * @param acknowledged The new value to assign to the alarm's acknowledged
+     *                     field.
+     */
+    public void setAcknowledged(final int id, final boolean acknowledged) throws IllegalArgumentException {
+        final Alarm alarm = getAlarmData(id);
+        if (alarm == null) {
+            throw new IllegalArgumentException("There is no alarm with ID " + id + ".");
+        }
+        alarm.setAcknowledged(acknowledged);
+    }
+
+    /**
+     * Raises an alarm with the given info, unless an alarm with the exact same
+     * parameters is already active.
+     */
+    private void raiseAlarm(
+        final String systemId,
+        final String systemName,
+        final Map<String, String> metadata,
+        final AlarmCause cause
+    ) {
+        synchronized (this) {
```

Comment:
You should avoid synchronized on this. Instead, you can use a member for this purpose:
`private final Object lock = new Object();`
See details in https://stackoverflow.com/questions/3369287/what-is-the-difference-between-synchronized-on-lockobject-and-using-this-as-the

Reply:
Good point, fixed.